README.md
# KaBoom - A High Performance Consumer Client for Kafka
KaBoom uses Krackle to consume from partitions of topics in Kafka and write them to boom files in HDFS.  

## Features
* Uses the [Curator Framework](http://curator.apache.org/) for  [Apache Zookeeper](zookeeper.apache.org) to distribute work amongst multiple servers
* Supports writing to secured Hadoop clusters via Kerberos based secure impersonation (conveniently pulled from [Flume](http://flume.apache.org/))
* Recovers from Kafka server failures (even when newly elected leaders weren't in-sync when elected)
* Supports consuming with either GZIP or Snappy compression
* Configurable: Each topic can be configured with a unique HDFS path template with date/time variable substitution
* Supports flagging timestamp template HDFS directories as 'Ready' when all a topic's partition's messages have been written for a given hour

## Author(s)
* [Dave Ariens](<mailto:dariens@blackberry.com>) ([Github](https://github.com/ariens))
* [Matthew Bruce](<mailto:mbruce@blackberry.com>) ([Github](https://github.com/MatthewRBruce))

## Building
Performing a Maven install produces: 

* An RPM package that currently installs RPM based Linux distributions
* A Debian package for dpkg based Linux distributions

## Major Changes in 0.8.x
This release contains the most significant updates to KaBoom we have introduced in a single version bump.  The most significant changes include the migration of all running configuration parameters and topic configurations to ZooKeeper and the introduction of worker sprints.  The remaining confiugration continues to be read in via a property file and is now referred to as startup configuration.  

## Worker Sprints

Workers are now aligned to sprints lasting `workerSprintDurationSeconds` seconds.  A sprint is associated to a partition ID (topic-partition), and starting Kafka offset.  A KaBoom worker will keep track of the partition IDs latest offest and largest parsed message timestamp for the duration of it's sprint.  Following `fileCloseGraceTimeAfterExpiredMs` milliseconds after the end of a sprint, KaBoom will then close off any open boom files and record the sprint's latest offset and largest observed timestamp to ZooKeeper.  

## Startup versus Running Configurations
Startup configuration changes require a KaBoom service restart to be loaded, whereas the running configuration is reloaded by KaBoom as changes are made in ZooKeeper.  Updated running configuration values are then used as they are accessed by KaBoom.  For example you can change the number of HDFS replicas to store for boom files in Hadoop however it will not affect any open or previously closed files only files that are created after the new configuration has been loaded (as replicas are specified only when creating files).

## Topic Configurations
Unlike running configurations which are reloaded instantly topic configuration updates trigger all workers assigned to a partition of the topic to be gracefully shutdown (boom files closed, offsets, and offset timestamps  stored in ZK).  The KaBoom client will then detect and restart any gracefully shutdown workers.  Workers load their topic configuration when they are launched.

## Example Topic Configuration
The topic configurations are stored at `zk://<root>/kaboom/topics/<id>`, as:

```
{
        version: 1,
        id: "devtest-test3",
        hdfsRootDir: "/service/82/devtest/logs/%y%M%d/%H/devtest-test3",
        proxyUser: "dariens",
        defaultDirectory: "data",
        filterSet: [ ]
}
```

Note: The empty filterSet array is reserved for future to-be-implemented  use-cases.

## Startup Configuration
Example startup configuration (property file based):

```
######################
# KaBoom Configuration
######################

kaboom.id=1001
hadooop.fs.uri=hdfs://hadoop.company.com
#kaboom.weighting=<number> - the default is number of cores
kerberos.keytab=/opt/kaboom/config/kaboom.keytab
kerberos.principal=kaboom@COMPANY.COM
#kaboom.hostname=<name> - the default is the system's hostname
zookeeper.connection.string=r3k1.kafka.company.com:2181,r3k2.kafka.company.com:2181,r3k3.kafka.company.com:2181/KaBoomDev
kafka.zookeeper.connection.string=r3k1.kafka.company.com:2181,r3k2.kafka.company.com:2181,r3k3.kafka.company.com:2181
#kaboom.load.balancer.type=even - this is the default
#kaboom.runningConfig.zkPath=/kaboom/config - this is the default

########################
# Consumer Configuration 
########################

metadata.broker.list=r3k1.kafka.company.com:9092,r3k2.kafka.company.com:9092,r3k3.kafka.company.com:9092
fetch.message.max.bytes=10485760
fetch.wait.max.ms=5000
#fetch.min.bytes=1 - this is the default
socket.receive.buffer.bytes=10485760
auto.offset.reset=smallest
#socket.timeout.seconds=30000 - this is the default
```

## Running Configuration
Here is an example running configuration stored at `zk://<root>/kaboom/config`:

```
{
        version: 8,
        allowOffsetOverrides: true,
        sinkToHighWatermark: true,
        useTempOpenFileDirectory: false,
        useNativeCompression: false,
        readyFlagPrevHoursCheck: 24,
        leaderSleepDurationMs: 600001,
        compressionLevel: 6,
        boomFileBufferSize: 16384,
        boomFileReplicas: 3,
        boomFileBlocksize: 268435456,
        boomFileTmpPrefix: "_tmp_",
        periodicHdfsFlushInterval: 30000,
        kaboomServerSleepDurationMs: 10000,
        fileCloseGraceTimeAfterExpiredMs: 30000,
        forcedZkOffsetTsUpdateMs: 600000,
        kafkaReadyFlagFilename: "_READY",
        maxOpenBoomFilesPerPartition: 5,
        workerSprintDurationSeconds: 3600,
        propagateReadyFlags: true,
        propagateReadyFlagFrequency: 600000,
        propateReadyFlagDelayBetweenPathsMs: 10
}
```

## Example Configuration File: /opt/kaboom/config/kaboom-env.sh
Here is an exmaple environemnt configuration file (defines runtime configuration and JVM properties):

```
JAVA=`which java`
BASEDIR=/opt/kaboom
BINDIR="$BASEDIR/bin"
LIBDIR="$BASEDIR/lib"
LOGDIR="/var/log/kaboom"
CONFIGDIR="$BASEDIR/config"
JMXPORT=9580
LOG4JPROPERTIES=$CONFIGDIR/log4j2.xml
PIDBASE=/var/run/kaboom
KABOOM_USER=kafka

JAVA_OPTS=""
JAVA_OPTS="$JAVA_OPTS -server"
JAVA_OPTS="$JAVA_OPTS -Xms6G -Xmx6G"
JAVA_OPTS="$JAVA_OPTS -XX:+UseParNewGC -XX:+UseConcMarkSweepGC"
JAVA_OPTS="$JAVA_OPTS -XX:+UseCMSInitiatingOccupancyOnly -XX:+CMSConcurrentMTEnabled -XX:+CMSScavengeBeforeRemark"
JAVA_OPTS="$JAVA_OPTS -XX:CMSInitiatingOccupancyFraction=30"

JAVA_OPTS="$JAVA_OPTS -XX:+PrintGCDetails -XX:+PrintGCDateStamps -XX:+PrintTenuringDistribution"
JAVA_OPTS="$JAVA_OPTS -Xloggc:$LOGDIR/gc.log -XX:+UseGCLogFileRotation -XX:NumberOfGCLogFiles=10 -XX:GCLogFileSize=10M"

JAVA_OPTS="$JAVA_OPTS -Djava.awt.headless=true"
JAVA_OPTS="$JAVA_OPTS -Dcom.sun.management.jmxremote"
JAVA_OPTS="$JAVA_OPTS -Dcom.sun.management.jmxremote.authenticate=false"
JAVA_OPTS="$JAVA_OPTS -Dcom.sun.management.jmxremote.ssl=false"
JAVA_OPTS="$JAVA_OPTS -Dcom.sun.management.jmxremote.port=$JMXPORT"

JAVA_OPTS="$JAVA_OPTS -Dlog4j.configurationFile=file:$LOG4JPROPERTIES"

JAVA_OPTS="$JAVA_OPTS -Dkaboom.logs.dir=$LOGDIR"

CLASSPATH=$CONFIGDIR:/etc/hadoop/conf:$LIBDIR/*
```

## Example Configuration File: /opt/kaboom/config/log4j2.xml (logging)
```
<?xml version="1.0" encoding="UTF-8"?>
<!-- This status="$LEVEL" on the next line  is for the logging of log4j2 as it configured tself, don't adjust it for  your application logging -->
<configuration status="WARN" monitorInterval="30">
   <appenders>
      <Console name="console" target="SYSTEM_OUT">
         <PatternLayout pattern="[%d] %p %m (%c)%n" />
      </Console>
      <RollingFile name="primary" fileName="/var/log/kaboom/server.log" filePattern="/var/log/kaboom/server.log.%d{yyyy-MM-dd-k}.log">
         <PatternLayout>
            <Pattern>[%d] %p %m (%c)%n</Pattern>
         </PatternLayout>
         <Policies>
            <TimeBasedTriggeringPolicy interval="1" modulate="true" />
         </Policies>
      </RollingFile>
   </appenders>
   <Loggers>
      <Logger name="stdout" level="info" additivity="false">
         <AppenderRef ref="console" />
      </Logger>
      <Root level="INFO">
         <AppenderRef ref="primary" />
      </Root>
   </Loggers>
</configuration>
```

## Running
After configuration simply start the kaboom service 'service kabom start'.

## Monitoring
Exposed via [Dropwizard Metric's](https://github.com/dropwizard/metrics) are metrics for monitoring message count, size, and lag (measure of how far behind KaBoom is compared to most recent message in Kafka--both in offset count and seconds):

New monitoring metrics in 0.7.1:

* Meter: boom writes (The number of boom file writes)

Kaboom (Aggregate metrics--for the KaBoom cluster):

* Gauge: max message lab sec 
* Gauge: sum message lag sec 
* Gauge: avg message lag sec 
* Gauge: max message lag 
* Gauge: sum message lag
* Gauge: avg message lag 
* Gauge: avg messages written per sec
* Gauge: total messages written per sec

Kaboom (Instance metrics -- for a KaBoom worker assigned to a topic and partition):

* Gauge: offset lag
* Gauge: seconds lag
* Gauge: messages written per second
* Gauge: early offsets received (when compression is enabled and messages are included from earlier than requested offset)
* Meter: boom writes

Krackle:

* Meter: MessageRequests
* Meter: MessageRequestsTotal
* Meter: MessagesReturned
* Meter: MessagesReturnedTotal
* Meter: BytesReturned
* Meter: BytesReturnedTotal
* Meter: MessageRequestsNoData
* Meter: MessageRequestsNoDataTotal
* Meter: BrokerReadAttempts
* Meter: BrokerReadAttemptsTotal
* Meter: BrokerReadSuccess
* Meter: BrokerReadSuccessTotal
* Meter: BrokerReadFailure
* Meter: BrokerReadFailureTotal

## Boom Files
This section contains portions from the [hadoop-logdriver](https://github.com/blackberry/hadoop-logdriver) project's description of Boom files.

A Boom file is a place where we store logs in HDFS.

The goals of Boom are:

* Be splittable by Hadoop, so that we can efficiently run MapReduce jobs against it.
* Be compressed to save storage.
* Be able to determine order of lines, even if they are processed out of order.

## File extention
The .bm file extension is used for Boom files.

## Boom File Format
A Boom file is a specific type of Avro [Object Container File](http://avro.apache.org/docs/1.6.3/spec.html#Object+Container+Files).  Familiarize yourself with those docs before you keep going.

Specifically, we always use a compression codec of 'deflate' and we always use the following Schema:

    {
      "type": "record",
      "name": "logBlock",
      "fields": [
        { "name": "second",      "type": "long" },
        { "name": "createTime",  "type": "long" },
        { "name": "blockNumber", "type": "long" },
        { "name": "logLines", "type": {
          "type": "array",
            "items": {
              "type": "record",
              "name": "messageWithMillis",
              "fields": [ 
                { "name": "ms",      "type": "long" },
                { "name": "message", "type": "string" }
              ]
            }
        }}
      ]
    }

### Basic Structure
The file contains any number of "logBlock" records.  Each logBlock contains data for multiple log lines, but all of the log lines in the record are timestamped in the same second.  Log lines in the same logBlock can have difference millisecond timestamps.

### Fields in logBlock
* second : the number of seconds since Jan 1, 1970 UTC.  All log lines in this record are timestamped with a time that occurs within this second.
* createTime : the time (in milliseconds) that this logBlock was created.  This is used for sorting logBlocks.
* blockNumber : a number indicating the sequence in which the logBlocks were written by whatever wrote the file.  This is used for sorting logBlocks.
* logLines : an array of "messageWithMillis" records, one per log line.

### Fields in messageWithMillis
* ms : the milliseconds part of the timestamp for this log line.  To get the complete timestamp, use second * 1000 + ms.
* eventId : an event identifier, reserved for future use.  Use 0 for raw log lines.
* message : the contents of the log line, excluding the timestamp and one space after the timestamp.

## Boom suggested defaults
Although no limitations should be assumed on the file beyond what has already been stated, these are sensible defaults that should be followed.

* The logLines field should contain no more that 1000 messageWithMillis entries.  If there are more than 1000 log lines within a second, then use multiple logBlock's with the same second value.
* The Avro Object Container File defines a "sync interval".  A good value for this seems to be 2MB (2147483648).
* While we are required to use the deflate codec, the compression level is configurable.  If you don't have a specific need, then level 6 is a good default.

## Sorting log lines
If the order of log lines is important, then the fields can be sorted by comparing fields in this order

* timestamp : first timestamp is first (after adding seconds and milliseconds)
* createTime : logBlocks that were written first go first.
* blockNumber : If two logBlocks were written in the same millisecond, then use them in the order they were written.
* index within logLines : If the log lines are the same timestamp, written in the same block, then the order is determined by where they are within the logLines array.

This is the default sorting for LogLineData objects.

## Contributing
To contribute code to this repository you must be [signed up as an official contributor](http://blackberry.github.com/howToContribute.html).

## Disclaimer
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


docs/changes.md
# KaBoom Changes

## 0.9.1-HF2

* KABOOM-47: Release 0.9.1-HF2 (pull in formal Krackle 0.9.2 release)

## 0.9.1-HF1

* KABOOM-46: Do not share consumer configurations

## 0.9.1

* KABOOM-44: Upgrade Hadoop Client to 2.7.2 and Login via Keytab

## 0.9.0

* KABOOM-3: Add JMX metric: Failed Block Writes
* KABOOM-31: expose bytes written to HDFS via JMX
* KABOOM-35: add a warning/error message in kaboom when /etc/hadoop/conf is not found or not defined in the classpath
* KABOOM-36: Expose login context name as a confguration option
* KABOOM-39: deprecate kerberos configs in kaboom.properties for kaboom versions >= 0.9.0
* KABOOM-40: Create KaBoom temp dir on package install and set snappy temp dir to use it
* KABOOM-41: Update Krackle dependency to 0.9.0 and remove -SNAPSHOT tag of BDP Common Dependency

## 0.8.4-HF2

* KABOOM-37: Old timestamps are not being treated as skewed

## 0.8.4-HF1

* KABOOM-33: Provide Better Resiliency around ZK Corruption

## 0.8.4

Tickets:

* KABOOM-26 Fix Overburdening Assignment (Even Load Balancer will not assign a partition to a node if that would make the node over-burdened).  EvenLoadBalancer ensures clients meet the criteria of canTakeAnotherAssignment() and  hasLocalPartition(KaBoomPartition partition) before assigning an unassigned partition
* KABOOM-27 Implement Skewed Timestamp (Old/Future Parsed Dates) Handling
* KABOOM-28 Improve Assignment and Lock Error Handling

Features:

TimeBasedHdfsOutputPath: New method skewed() return true if the boom files date directory is too far into the past/future based on new running configuration options
TimeBasedHdfsOutputPath: OutputFile private class overwrites the boom filename, date directory, and data directory according to new running configuration options

New Metrics:

* kaboom:total:skewed time boom files // the total number of skewed boom files for the node
* kaboom:partitions:<partition>:skewed time boom files // the total number of skewed boom files for the partiton

## 0.8.3 (never released outside of labs)

* kaboom-api dependency now on version 0.8.4
* AsyncAssignee: moved node cache listener on assignment path to attribute and closes on release lock
* AsyncAssignee: moved connection state listener to attribute and removes on release lock
* TimeBasedHdfsOutputPath now waits for existing open files to close waiting nodeOpenFileWaittimeMs between isFileClosed()
* TimeBasedHdfsOutputPath gives up and deletes existing open files if not closed after nodeOpenFileForceDeleteSeconds
* TimeBasedHdfsOutputPath now has an instance of the Worker as an attribute and can respond to pings when it's busy
* Leader has a new method called refreshMetadata() that clears all convenience mappings and refreshes all metadata
* Leader has a new method called pauseOnFirstDisconnectedAssignee() that iterates through all assignments and waits leaderNodeDisconnectionWaittimeSeconds on the first disconnected assigned node before calling refreshMetadata() and returning
* Leader's main loop now calls refreshMetadata() and then pauseOnFirstDisconnectedAssignee()
* Leader worst case running time increased by at least leaderNodeDisconnectionWaittimeSeconds


## 0.8.2-HF4

* Updates the kaboom-api dependency to 0.8.3 (which resolves leaking Kafka simple consumer sockets)
* Exposes the startup config's node cache listener as an accessible attribute
* Removes unnecessary and unused mapping of proxy user to file system

## 0.8.2-HF3

* Fixes KABOOM-20 - Maps Used by Leader Not Emptied
* Fixes KABOOM-21 - Ensure Exceptions Are Not Swallowed And Abort Accordingly
* Fixes KABOOM-22 - Bug in Closing LRU Boom File Closing Most Recently Used v2

## 0.8.2-HF2

* Fixes IPGBD-4245/KABOOM-18 - Bug closing LRU Output File


## 0.8.2-HF1

* Moves the node cache listener to an attribute of worker and closes off in the finally block executed after the run

## 0.8.2:

* TODO: Re-write the change log

## 0.8.1

* New metric, kaboom:total:gracefully restarted workers

## 0.8.0

* Refactored configuration into startup/running, running configuration migrated to ZK
* Introduction of worker sprints
* Intended to be manage via Kontroller API and web interface

## 0.7.16-HF1

* Improves exception handling and reduce file corruption when closing boom files to ensure that any problems closing the file result in the worker aborting the file.

## 0.7.16

* Adds a new stand-alone utility to write a boom file for a specific partition, start offset and end offset to a specific destination
* Adds a new meter metric: kaboom:total:dead workers

## 0.7.15-HF1

* Fixes IPGBD-3830 [Kaboom] Bug when handling zero padded date field

## 0.7.15

* Ensures that only topics with unique HDFS root paths are examined
* New configuration option: kaboom.propagate.ready.flags.delay.ms (long, how often wait between paths, default 0) can be used to ease the burden on the name nodes if there's a massive amount of HDFS path traversal required

## 0.7.14

* Moved the READY flag propagation to KaBoom (from an older to-be-deprecated project
* New JMX metric: kaboom:topic:<topicName>:flag propagator timer (timer for recursive HDFS directory traversal)
* New JMX metric: kaboom:topic:<topicName>:flag propagator paths checked (merter for number of paths checked)
* New JMX metric: kaboom:topic:<topicName>:flags written (merter for number of flags created)
* New configuration option: kaboom.propagate.ready.flags (boolean, default false)
* New configuration option: kaboom.propagate.ready.flags.frequency (long, how often in ms to spawn propagator thread, default 10 * 60 * 1000)
* Supports bdp-common 0,0.6 which provides all logging and monitoring deps
* Instruments log4j2 with io.dropwizard.metrics
* Adds a new API

## 0.7.13

* Added kafkaOffset argument to FastBoomWriter.writeLine()
* Added a lastKafkaOffset to FastBoomWriter
* FastBoomWriter now keeps track of lastKafkaOffset and lastMessageTimestamp

conf/kaboom-env.sample.sh
#!/bin/bash

JAVA='which java'
BASEDIR=/opt/kaboom
BINDIR="$BASEDIR/bin"
LIBDIR="$BASEDIR/lib"
LOGDIR="/var/log/kaboom"
CONFIGDIR="$BASEDIR/config"
JMXPORT=9580
LOG4JPROPERTIES=$CONFIGDIR/log4j.properties
PIDBASE=/var/run/kaboom
KABOOM_USER=kafka

JAVA_OPTS=""
JAVA_OPTS="$JAVA_OPTS -server"
JAVA_OPTS="$JAVA_OPTS -Xms10G -Xmx10G"
JAVA_OPTS="$JAVA_OPTS -XX:+UseParNewGC -XX:+UseConcMarkSweepGC"
JAVA_OPTS="$JAVA_OPTS -XX:+UseCMSInitiatingOccupancyOnly -XX:+CMSConcurrentMTEnabled -XX:+CMSScavengeBeforeRemark"
JAVA_OPTS="$JAVA_OPTS -XX:CMSInitiatingOccupancyFraction=30"

JAVA_OPTS="$JAVA_OPTS -XX:+PrintGCDetails -XX:+PrintGCDateStamps -XX:+PrintTenuringDistribution"
JAVA_OPTS="$JAVA_OPTS -Xloggc:$LOGDIR/gc.log -XX:+UseGCLogFileRotation -XX:NumberOfGCLogFiles=10 -XX:GCLogFileSize=10M"

JAVA_OPTS="$JAVA_OPTS -Djava.awt.headless=true"
JAVA_OPTS="$JAVA_OPTS -Dcom.sun.management.jmxremote"
JAVA_OPTS="$JAVA_OPTS -Dcom.sun.management.jmxremote.authenticate=false"
JAVA_OPTS="$JAVA_OPTS -Dcom.sun.management.jmxremote.ssl=false"
JAVA_OPTS="$JAVA_OPTS -Dcom.sun.management.jmxremote.port=$JMXPORT"

JAVA_OPTS="$JAVA_OPTS -Dlog4j.configuration=file:$LOG4JPROPERTIES"

JAVA_OPTS="$JAVA_OPTS -Dkaboom.logs.dir=$LOGDIR"

CLASSPATH=$CONFIGDIR:/etc/hadoop/conf:$LIBDIR/*